### PR TITLE
ARUHA-592: faceted metrics

### PR DESCRIPTION
**Pull request by rcillo**
https://techjira.zalando.net/browse/ARUHA-592

Aruha team used to have a hard time to quickly answer questions about specific application performance.

Questions like:

- Who consumes my event types and how much?
- How many active consumer a given application has at the moment?
- How fast are Kafka consumers streaming?

In order not to grow even more our existent `/metrics` endpoint, I'm created 2 new endpoints:

1. /requests-metrics
- request timer on a per application basis for every URI path:  application-id.http-method.http-path
- counter for number of consumers from a given subscription: nakadi.subscriptions.subscription-id.consumers
2. /stream-metrics
- it exposes Kafka consumer metrics for HiLA: hila.application-id.subscription-id.bytes-flushed
- it exposes Kafka consumer metrics for LoLA: lola.application-id.event-type.bytes-flushed

With all these metrics, we are going to be able to detect slowdowns specific to certain applications, subscriptions or event types. Also, we will be able to detect who's abusing the system, by calling a given endpoint too often or streaming too much.

All of this was only possible previously through log analyses.

### CODE REVIEW

File: src/main/java/org/zalando/nakadi/controller/EventStreamController.java [hunk 1]
```diff
@@ -174,6 +182,12 @@ public StreamingResponseBody streamEvents(
             final AtomicBoolean connectionReady = closedConnectionsCrutch.listenForConnectionClose(request);
             Counter consumerCounter = null;
             EventStream eventStream = null;
+            EventConsumer eventConsumer = null;
+            final String kafkaClientMetricsName = MetricRegistry.name(
+                    "lola",
+                    client.getClientId(),
+                    eventTypeName,
```
Comment (rcillo) [inline src/main/java/org/zalando/nakadi/controller/EventStreamController.java]: Maybe I should remove `.` from event type name?

File: src/main/java/org/zalando/nakadi/controller/SubscriptionStreamController.java [hunk 2]
```diff
@@ -124,12 +132,17 @@ public StreamingResponseBody streamEvents(
...
 
         return outputStream -> {
+            Counter consumerCounter = null;
 
             if (!featureToggleService.isFeatureEnabled(HIGH_LEVEL_API)) {
                 response.setStatus(HttpServletResponse.SC_NOT_IMPLEMENTED);
                 return;
             }
 
+            final String metricName = metricNameForSubscription(subscriptionId, CONSUMERS_COUNT_METRIC_NAME);
```
Comment (rcillo) [inline src/main/java/org/zalando/nakadi/controller/SubscriptionStreamController.java]: Maybe I should also scope this metric per `application-id`?

File: src/main/java/org/zalando/nakadi/controller/EventStreamController.java [hunk 3]
```diff
@@ -254,6 +273,10 @@ public StreamingResponseBody streamEvents(
                 if (eventStream != null) {
                     eventStream.close();
                 }
+                if (eventConsumer != null) {
+                    this.kafkaClientMetrics.removeMatching((name, metric) -> {
+                        return name.startsWith(kafkaClientMetricsName); });
```
Comment (adyach) [inline src/main/java/org/zalando/nakadi/controller/EventStreamController.java]: No need in `return`

File: src/main/java/org/zalando/nakadi/controller/EventStreamController.java (hunk 1 above)
Comment (adyach) [inline src/main/java/org/zalando/nakadi/controller/EventStreamController.java]: Do you consider using MetricUtils to generate the name? It replaces `.` in event type name with `#`.

File: src/main/java/org/zalando/nakadi/metrics/KafkaClientMetrics.java [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+    public KafkaClientMetrics(final Consumer<String, String> kafkaConsumer) {
+        this.kafkaConsumer = kafkaConsumer;
+    }
+
+    @Override
+    public Map<String, Metric> getMetrics() {
+        final Map<String, Metric> gauges = new HashMap<>();
+        for (final Object metricObject : kafkaConsumer.metrics().entrySet()) {
+            final Map.Entry<MetricName, KafkaMetric> metricEntry = (Map.Entry<MetricName, KafkaMetric>) metricObject;
+            gauges.put(metricEntry.getKey().name(), (Gauge) () -> { return metricEntry.getValue().value(); });
```
Comment (adyach) [inline src/main/java/org/zalando/nakadi/metrics/KafkaClientMetrics.java]: No need in `return`

File: src/main/java/org/zalando/nakadi/metrics/KafkaClientMetrics.java [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+    private final Consumer<String, String> kafkaConsumer;
+
+    public KafkaClientMetrics(final Consumer<String, String> kafkaConsumer) {
+        this.kafkaConsumer = kafkaConsumer;
+    }
+
+    @Override
+    public Map<String, Metric> getMetrics() {
+        final Map<String, Metric> gauges = new HashMap<>();
+        for (final Object metricObject : kafkaConsumer.metrics().entrySet()) {
```
Comment (adyach) [inline src/main/java/org/zalando/nakadi/metrics/KafkaClientMetrics.java]: You could use Map.Entry<MetricName, ? extends org.apache.kafka.common.Metric>

File: src/main/java/org/zalando/nakadi/controller/SubscriptionStreamController.java (hunk 2 above)
Comment (adyach) [inline src/main/java/org/zalando/nakadi/controller/SubscriptionStreamController.java]: I think we can add it in the feature, if there is a need.

File: src/main/java/org/zalando/nakadi/controller/SubscriptionStreamController.java [hunk 6]
```diff
@@ -124,12 +132,17 @@ public StreamingResponseBody streamEvents(
             throws IOException {
 
         return outputStream -> {
+            Counter consumerCounter = null;
```
Comment (adyach) [inline src/main/java/org/zalando/nakadi/controller/SubscriptionStreamController.java]: You can move it to the metric construction (down there) and make final.

File: src/main/java/org/zalando/nakadi/metrics/MonitoringRequestFilter.java [hunk 7]
```diff
@@ -34,11 +41,28 @@ public void doFilter(final ServletRequest request, final ServletResponse respons
         openHttpConnectionsCounter.inc();
         final Timer.Context timerContext = httpConnectionsTimer.time();
 
+        Timer.Context perPathTimerContext = null;
+        if (request instanceof HttpServletRequest) {
+            final HttpServletRequest httpRequest = (HttpServletRequest) request;
+            final String clientId = Optional.ofNullable(httpRequest.getUserPrincipal())
+                    .map(Principal::getName)
+                    .orElse("unauthenticated");
```
Comment (adyach) [inline src/main/java/org/zalando/nakadi/metrics/MonitoringRequestFilter.java]: Is it useful to gather metrics for unauthenticated users? and what can we gather?

**Comment by adyach** [comment]
It looks good 👍 except tests are not passing yet and some minor things I've noticed

File: src/main/java/org/zalando/nakadi/metrics/MonitoringRequestFilter.java (hunk 7 above)
Comment (rcillo) [inline src/main/java/org/zalando/nakadi/metrics/MonitoringRequestFilter.java]: for example, `/health` and `/metrics` endpoint. We might realise that they are called too often for example.

**Comment by rcillo** [comment]
@adyach fixed comments and simplified metrics. Now there is going to be a `/stream-metrics` endpoint with stream metrics. It's not bound to Kafka anymore.

File: src/main/java/org/zalando/nakadi/controller/EventStreamController.java [hunk 8]
```diff
@@ -219,8 +225,18 @@ public StreamingResponseBody streamEvents(
                 final EventConsumer eventConsumer = topicRepository.createEventConsumer(
                         kafkaQuotaClientId,
                         streamConfig.getCursors());
+
+                final String bytesFlushedMetricName = MetricRegistry.name(
```
Comment (adyach) [inline src/main/java/org/zalando/nakadi/controller/EventStreamController.java]: Do you think it makes sense to use it as in subscriptions through MetricUtils?

Comment (rcillo) [inline src/main/java/org/zalando/nakadi/controller/EventStreamController.java]: This name looks like "lola.stups_some-application-name.some#event-ype-name.bytes-flushed". The existent methods in `MetricUtils` looks like "nakadi.subscriptions.subscription-id.bytes-flushed". It has 2 problems:

1. it has no space for application-id
2. it has no space for lola/hila.

I could create new methods there, to have it all in a single class, though. Is this what you meant?

**Comment by adyach** [comment]
👍

File: src/main/java/org/zalando/nakadi/controller/EventStreamController.java (hunk 8 above)
Comment (rcillo) [inline src/main/java/org/zalando/nakadi/controller/EventStreamController.java]: @adyach done 92cc082

**Comment by rcillo** [comment]
@adyach I need a new approval after I refactored `MetricUtils` a little bit.

**Comment by adyach** [comment]
👍

**Comment by rcillo** [comment]
@adyach I used spring filters and it seems to have solved. I tried locally with authentication turned on.

**Comment by rcillo** [comment]
the thing about it is that it no longer registers the time taken to authenticate, obviously.

File: src/main/java/org/zalando/nakadi/metrics/MonitoringRequestFilter.java [hunk 9]
```diff
@@ -3,61 +3,56 @@
...
-        }
+        final Timer.Context perPathTimerContext = perPathMetricRegistry.timer(perPathMetricKey).time();
 
         try {
 
-            chain.doFilter(request, response);
+            filterChain.doFilter(request, response);
 
         } finally {
             if (perPathTimerContext != null) {
```
Comment (adyach) [inline src/main/java/org/zalando/nakadi/metrics/MonitoringRequestFilter.java]: Is it possible that it is `null` here?

Comment (rcillo) [inline src/main/java/org/zalando/nakadi/metrics/MonitoringRequestFilter.java]: Not at all. I forgot to remove it after refactoring. Done https://github.com/zalando/nakadi/pull/578/commits/c06e4871d24dd3e67f4a6d41f22f5b01ab27b1b7

**Comment by adyach** [comment]
👍

**Comment by adyach** [comment]
:shipit:

**Pull request merged**